specification/dsdl/serializable_types.tex

```diff
@@ -0,0 +1,285 @@
+at the data type definition time,
+in which case the array is said to be a \emph{variable-length array}.
+Variable-length arrays with unbounded maximum number of elements are not allowed.
+
+Arrays are defined by adding a pair of square brackets after the array element type specification,
+where the brackets contain the \emph{array capacity expression}.
+The array capacity expression must yield a positive integer upon its evaluation;
+any other value or type renders the current DSDL definition invalid.
+
+The array capacity expression can be prefixed with the following character sequences in order to define
+a variable-length array:
+\begin{itemize}
+    \item ``\verb|<|'' (ASCII code 60) --- indicates that the integer value yielded by the array capacity expression
+    specifies the non-inclusive upper boundary of the number of elements.
+
+    \item ``\verb|<=|'' (ASCII code 60 followed by 61) --- same as above, but the upper boundary is inclusive.
+\end{itemize}
+If neither of the above prefixes are provided, the resultant definition is that of a fixed-length array.
+
+\subsection{Composite types}\label{sec:dsdl_composite_types}
+
+\subsubsection{Kinds}
+
+There are two kinds of composite data type definitions: message types and service types.
+All versions of a data type must be of the same kind\footnote{%
+For example, if a data type version 0.1 is of a message kind, all later versions of it must be messages, too.}.
+
+A message data type defines one data schema.
+
+A service data type contains two data schema definitions:
+one for service request object, and one for service response object, in that order.
+The two schemas are separated by the service response marker (``\verb|---|'') on a separate line.
+
+The presence of the service response marker indicates that the data type definition at hand is of the service kind.
+There may be no more than one service response marker in a given definition.
+
+\subsubsection{Dependencies}
+
+In order to refer to a composite type from another composite type definition
+(e.g., for nesting or for referring to an external constant),
+one has to specify the full data type name of the referred data type followed by its
+major and minor version numbers separated by the namespace separator character,
+as demonstrated on the figure \ref{fig:dsdl_nested_reference}.
+
+To facilitate look-up of external dependencies,
+implementations are expected to obtain from external sources%
+\footnote{For example, from user-provided configuration options.}
+the list of directories that are the roots of namespaces containing the referred dependencies.
+
+\begin{figure}[H]
+    $$
+    \underbrace{\texttt{\huge{uavcan.node.Heartbeat}}}_{\text{full data type name}}%
+    \texttt{\huge{.}}%
+    \underbrace{\texttt{\huge{1.0}}}_{\substack{\text{version} \\ \text{numbers}}}%
+    $$
+    \caption{Reference to an external composite data type definition.\label{fig:dsdl_nested_reference}}
```

Comment:
Should we support a >=< syntax to allow some flexibility here? We might otherwise cause cascading changes that were unnecessary.

Reply:
Would be a mistake. See this and the next two posts: https://forum.uavcan.org/t/the-case-for-code-compatibility/194/10?u=pavel.kirienko

Currently, we provide a hard guarantee that once a given version of a data type is released, it is to be never changed again excepting whitespaces, comments, and `@deprecated`. It is important for maintainability and compatibility. Automatic switch to a newer version would remove this guarantee.